### DISCUSSION TITLE
Action function stream response

### DISCUSSION THREAD

**Discussion by rreeves8** [remix-run/remix#8695] [Q&A] [answered]
Hi, I want to stream a bunch of text data back to the client from an action function. 

<img width="297" alt="Screenshot 2024-02-06 at 3 58 42 PM" src="https://github.com/remix-run/remix/assets/71095829/d3e1f49b-2d84-4771-a345-32406e3539e7">

Use action data only returns the data after the stream is closed. Is there a way to get each update from the stream ?

<img width="479" alt="Screenshot 2024-02-06 at 3 59 35 PM" src="https://github.com/remix-run/remix/assets/71095829/dfefadf2-d117-4baa-afa8-4219ca2cfa5e">

^ test code

**Comment by gustavopch** [comment]
I had a similar problem on Firebase Functions. It was buffering the stream (it was actually Firebase Hosting's CDN doing that) and only delivering after finished. Had to move to Fly to be able to stream properly.

**Reply by rreeves8** [reply]
How did you make the client hook update the document on every stream event ? Was it the response headers ?

**Reply by gustavopch** [reply]
I'm just using `defer`, not dealing with the stream directly, so not entirely sure if the problem you're having is the same as mine.

**Comment by ganesh-rao** [comment]
I'm facing the same scenario. Have you had any luck with resolving this?

**Comment by kiliman** [comment]
Here's an article on how to use event streams with Remix.

https://sergiodxa.com/tutorials/use-server-sent-events-with-remix

I also have an example here. The event stream example does not depend on Single Data Fetch, so it still applies.

https://github.com/kiliman/remix-single-fetch

**Comment by ganesh-rao** [comment]
@kiliman Thanks for your reply. I don't believe that is the solution to the issue raised in this thread as SSE works only for GET requests. The action function is invoked on POST requests.

If anyone has any other ideas, please share!

**Reply by rreeves8** [reply]
Fr, the ability to return a readable stream would be nice.

**Reply by vladinator1000** [reply]
Bummer that Remix doesn't support this. Would love a framework-blessed way to do this:

```tsx
// some-action.ts
export async function action({ request, params, context }: Route.ActionArgs) {
  // ...
  return stream
}

export default Component() {
  let fetcher = useStreamingFetcher()
  let [streamOutput, setStreamOutput] = useState('')
  
  useEffect(() => {
    if (fetcher.message === "message") {
      setStreamOutput((old) => old + fetcher.data)
    } 
  }, [fetcher.message, fetcher.data])
  
  return <fetcher.Form
    action="some-action"
    encType="multipart/form-data"
    method="post"
  >
    <input type="file" name="file" />
  </fetcher.Form>
}
```

**Comment by admirsaheta** [comment, marked as answer]
Streaming text data from an action function can get tricky,  Remix doesn't have built-in functionality for it. But, we can still make it work!

Might be https://remix.run/docs/en/main/guides/streaming

WebSockets are best for real-time, back-and-forth communication.  For pushing data from server to client,  SSE or Action Data Chunking could work depending on your needs.

Action Data Chunking  isn't true streaming, but it can feel similar. We'd break down your data into smaller portions.  In your action function, process the data chunk by chunk. After each chunk is done, you'd return an object containing the processed data and a flag indicating if there's more data to come. The client-side would handle the response and keep making requests until the flag says "stop, we're done!"

Let me know if you have any questions about these approaches, or if there's another way you were thinking of tackling this.

**Reply by lodi-g** [reply]
Hey @admirsaheta,

Thanks for the tips. I think WebSockets are a bit overkill for what I'm trying to build.
Can you expand on Action Data Chunking? What did you mean?

Thanks!
